objects/src/accounts/delta/mod.rs

```diff
@@ -0,0 +1,186 @@
+///
+/// TODO: add ability to trace account code updates.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct AccountDelta {
+    storage: AccountStorageDelta,
+    vault: AccountVaultDelta,
+    nonce: Option<Felt>,
+}
+
+impl AccountDelta {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns new [AccountDelta] instantiated from the provided components.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Storage or vault deltas are invalid.
+    /// - Storage and vault deltas are empty, and the nonce was updated.
+    /// - Storage or vault deltas are not empty, but nonce was not updated.
+    pub fn new(
+        storage: AccountStorageDelta,
+        vault: AccountVaultDelta,
+        nonce: Option<Felt>,
+    ) -> Result<Self, AccountDeltaError> {
+        // make sure storage and vault deltas are valid
+        storage.validate()?;
+        vault.validate()?;
+
+        // nonce must be updated if and only if either account storage or vault were updated
+        validate_nonce(nonce, &storage, &vault)?;
+
+        Ok(Self {
+            storage,
+            vault,
+            nonce,
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns true if this account delta does not contain any updates.
+    pub fn is_empty(&self) -> bool {
+        self.storage.is_empty() && self.vault.is_empty()
+    }
+
+    /// Returns storage updates for this account delta.
+    pub fn storage(&self) -> &AccountStorageDelta {
+        &self.storage
+    }
+
+    /// Returns vault updates for this account delta.
+    pub fn vault(&self) -> &AccountVaultDelta {
+        &self.vault
+    }
+
+    /// Returns the new nonce, if the nonce was changes.
+    pub fn nonce(&self) -> Option<Felt> {
+        self.nonce
+    }
+
+    /// Converts this storage delta into individual delta components.
+    pub fn into_parts(self) -> (AccountStorageDelta, AccountVaultDelta, Option<Felt>) {
+        (self.storage, self.vault, self.nonce)
+    }
+}
+
+impl Serializable for AccountDelta {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        self.storage.write_into(target);
+        self.vault.write_into(target);
+        self.nonce.write_into(target);
+    }
+}
+
+impl Deserializable for AccountDelta {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let storage = AccountStorageDelta::read_from(source)?;
+        let vault = AccountVaultDelta::read_from(source)?;
+        let nonce = <Option<Felt>>::read_from(source)?;
+
+        validate_nonce(nonce, &storage, &vault)
+            .map_err(|err| DeserializationError::InvalidValue(err.to_string()))?;
+
+        Ok(Self {
+            storage,
+            vault,
+            nonce,
+        })
```

Comment:
I thought about this, but then eventually decided to go with this approach because I think of deserialization as a more adversarial process, and so, we wan'd want to fail as soon as an error is detected. Doing it via `new()` would require deserializing the entire struct first and then validating it.

The counter-argument to the above is that performance difference is probably negligible and we benefit a lot from having unified validation logic. I've tried to address this via tests. But I can also see how this may be a case of premature optimization.

Reply:
> The counter-argument to the above is that performance difference is probably negligible and we benefit a lot from having unified validation logic.

Hmm yeah probably. We're of course guessing because we have no data from production, but my guess is 99% of the time, there'll be no deserialization error.

In fact, I just noticed now from your comment that `AccountStorageDelta::read_from()` also performs validation (supposedly the same one from `AccountStorageDelta::validate()`), which is surprising to me. My assumption when reading this code was that we forgot to call `storage.validate()?` and `vault.validate()?`.